### DISCUSSION TITLE
用zig的for作为表达式，实现reduce的功能？

### DISCUSSION THREAD

**Discussion by 4IK1d** [zigcc/forum#135] [新手提问] [answered]
大佬们，如下写法实现reduce的功能是否是合理的？
```ziglang
var result += for (array) |value| {
    value;
} else 0;
```

看到https://course.ziglang.cc/basic/process_control/loop#%E4%BD%9C%E4%B8%BA%E8%A1%A8%E8%BE%BE%E5%BC%8F%E4%BD%BF%E7%94%A8
这里似乎只是做了遍历和条件选取，似乎不能做到累加这种？

**Comment by jiacai2050** [comment]
```
test "for else" {
    // For allows an else attached to it, the same as a while loop.
    const items = [_]?i32{ 3, 4, null, 5 };

    // For loops can also be used as expressions.
    // Similar to while loops, when you break from a for loop, the else branch is not evaluated.
    var sum: i32 = 0;
    const result = for (items) |value| {
        if (value != null) {
            sum += value.?;
        }
    } else blk: {
        try expect(sum == 12);
        break :blk sum;
    };
    try expect(result == 12);
}
```
累加要放到 for 里面，https://ziglang.org/documentation/master/#toc-for

**Reply by 4IK1d** [reply]
感谢回答，但疑问确实不是如何实现，而是是否可以考虑像上边那样直接用表达式实现。就算是语法糖也好，看起来会直白很多

**Reply by jiacai2050** [reply, marked as answer]
原来你是这个意思。

但觉得不是个好设计，你觉得栗子还比较简单，如果是个string，那你怎么处理？没有地方传 allocator 了，缺少一个 context。

**Reply by zhang-chuanfeng** [reply]
> 感谢回答，但疑问确实不是如何实现，而是是否可以考虑像上边那样直接用表达式实现。就算是语法糖也好，看起来会直白很多

按照正常理解也是  for先执行玩  后面才会 += 赋值变量result啊。+= 两边类似表达式值  两边值得到才进行 +=
